src/analyses/dirty.h

```diff
@@ -17,42 +17,74 @@ Date: March 2013
 #include <unordered_set>
 
 #include <util/std_expr.h>
+#include <util/invariant.h>
 #include <goto-programs/goto_functions.h>
 
 class dirtyt
 {
+private:
+  void die_if_uninitialized() const
+  {
+    INVARIANT(initialized,
+        "Uninitialized dirtyt. This dirtyt was constructed using the default "
+        "constructor and not subsequently initialized using "
+        "dirtyt::build().");
+  }
+
 public:
+  bool initialized;
   typedef std::unordered_set<irep_idt, irep_id_hash> id_sett;
   typedef goto_functionst::goto_functiont goto_functiont;
 
-  explicit dirtyt(const goto_functiont &goto_function)
+  dirtyt():
+    initialized(false)
+  {
+  }
+
+  explicit dirtyt(const goto_functiont &goto_function):
+    initialized(false)
   {
     build(goto_function);
+    initialized=true;
   }
 
-  explicit dirtyt(const goto_functionst &goto_functions)
+  explicit dirtyt(const goto_functionst &goto_functions):
+    initialized(false)
   {
-    forall_goto_functions(it, goto_functions)
-      build(it->second);
+    build(goto_functions);
+    // build(g_funs) responsible for setting initialized to true, since
```

Comment:
This possibly isn't your design - but it would seem to be better for this constructor only to be visible, then you don't need to separately track initialised and have constructors as the only way build this object (of course removing the default constructor).